[ACCUMULO-4787] Close input stream in AccumuloReplicaSystem

adamjshook opened the pull request:
Contemplating how to make a test for this, but pushing this build out to our dev system has removed the `CLOSE_WAIT` connections.

ctubbsii commented on server/tserver/src/main/java/org/apache/accumulo/tserver/replication/AccumuloReplicaSystem.java:
```diff
@@ -420,91 +420,99 @@ protected Status replicateLogs(ClientContext peerContext, final HostAndPort peer
+      }
+    } finally {
+      try {
+        input.close();
```
It might be better to allocate input in a try-with-resources block instead. That way, it is always closed, even if the method returns normally before it gets to this while(true) loop (I noticed another try-catch above which returns, and leaves the input unclosed).

ctubbsii commented on server/tserver/src/main/java/org/apache/accumulo/tserver/replication/AccumuloReplicaSystem.java:
```diff
@@ -420,91 +420,99 @@ protected Status replicateLogs(ClientContext peerContext, final HostAndPort peer
+      try {
+        input.close();
+      } catch (IOException e) {
+        log.warn("Received IOException trying to close input stream: {}", e);
```
This log message should either not include the `: {}` part, or it should be have the `e` parameter twice. As is, it will convert the exception to a string to substitute for `{}` and we'd lose the stack trace.

adamjshook commented:
@ctubbsii Thanks for the review!  Updated per your comments.

joshelser commented:
> Good find. I think we should apply this back to the 1.7 branch, and use try-with-resources to properly close the input stream under all cases.

+1 to that.

I'm curious: why did findbugs not catch this? Anyone look at that yet? I thought our use of findbugs was rather rigorous.

ctubbsii commented:
> why did findbugs not catch this?

Probably because the resource is being allocated by a method... like a factory pattern. Unclosed resources allocated that way are difficult to track. It probably shows up in Eclipse as "potential resource leak", but that warning is usually suppressed with default Eclipse settings, because it's pretty spammy and full of false positives, because Java does not have a good way to track resource from factories.

joshelser commented:
Ahh, the two-hop makes it hard to follow. Thanks, I didn't think about that.

ctubbsii commented:
@adamjshook Were there additional changes, unrelated to the original issue also done in your updated commits? It seems like a larger change was made, but I'm having trouble following it.

adamjshook commented:
@ctubbsii The `try-with-resources` caused an indentation change that makes it look larger and I moved a large block of code into a new method to make it more readable.  The summary of the changes are:

1. Added a `try-with-resources` in the [`replicateRFiles`](https://github.com/apache/accumulo/pull/369/files#diff-48d2b9734f0e624e93e136f24f35e0bbR324) and [`replicateLogs`](https://github.com/apache/accumulo/pull/369/files#diff-48d2b9734f0e624e93e136f24f35e0bbR369)
2. Moved [processing of the WAL header](https://github.com/apache/accumulo/pull/369/files#diff-48d2b9734f0e624e93e136f24f35e0bbL374) into [`getWalStream`](https://github.com/apache/accumulo/pull/369/files#diff-48d2b9734f0e624e93e136f24f35e0bbR685).
3. Moving the exceptions that would be thrown from `getInputStream` [here](https://github.com/apache/accumulo/pull/369/files#diff-48d2b9734f0e624e93e136f24f35e0bbL327) and [here](https://github.com/apache/accumulo/pull/369/files#diff-48d2b9734f0e624e93e136f24f35e0bbL378) down to the bottom of the `try-with-resources`.
4. Refactoring [this](https://github.com/apache/accumulo/blob/53c0e4ffd9083a7dc13c0d93c65b6a081da0cf56/server/tserver/src/main/java/org/apache/accumulo/tserver/replication/AccumuloReplicaSystem.java#L332) block of code into a [`processBatches`](https://github.com/apache/accumulo/pull/369/files#diff-48d2b9734f0e624e93e136f24f35e0bbR384) method.

ctubbsii commented:
Ah, it was the relocation to a new method that screwed me up. If you ignore whitespace, the regular diff without that change is much better. To ignore whitespace in the GitHub.com view, add `?w=1` to the end of the diff URL.

ctubbsii commented:
To do a diff on the command-line with git, just do: `git diff --ignore-all-space`

adamjshook commented:
>To ignore whitespace in the GitHub.com view, add ?w=1 to the end of the diff URL.

TIL 🙌

adamjshook commented:
Any more changes left for this or are we good to merge?

ctubbsii commented:
@adamjshook I'd like to update my review, but with the code relocations, I'm still having trouble following it. If you could update the code to undo the relocations, so a clean diff, ignoring whitespace, can be viewed, I could update my review.

keith-turner commented on server/tserver/src/main/java/org/apache/accumulo/tserver/replication/AccumuloReplicaSystem.java:
```diff
@@ -321,60 +321,67 @@ public String execute(ReplicationCoordinator.Client client) throws Exception {
-    span.data("file", p.toString());
-    try {
-      input = getWalStream(p);
+    try (final DataInputStream input = getWalStream(p)) {
```
The call to `getWalStream()` calls `DfsLogger.readHeaderAndReturnStream()` which calls `fs.open()`.  If there is IOException in `readHeaderAndReturnStream()` or in `getWalStream()` then is seems like the FSDataInputStream will not be closed.

Not sure what the best way to handle this is, but one possibility I thought of is to refactor the code to call `fs.open()` in this method and then pass that down.  Maybe something like

```java
try(FSDataInputStream fsdi = fs.open(path)){
   DataInputStream input = getWalStream(fsdi);
}
```

or could do the following, I think java will close things in the reverse order from opened.

```java
try(FSDataInputStream fsdi = fs.open(path); DataInputStream input = getWalStream(fsdi)){
 
}
```

adamjshook commented:
@ctubbsii @keith-turner Undo on the code relocations and applied Keith's suggestion to refactor out the `FSDataInputStream`.  Back to you!

joshelser commented on server/tserver/src/main/java/org/apache/accumulo/tserver/replication/AccumuloReplicaSystem.java:
```diff
@@ -370,11 +368,112 @@ protected Status replicateLogs(ClientContext peerContext, final HostAndPort peer
-    span.data("file", p.toString());
-    try {
-      input = getWalStream(p);
+    try (final FSDataInputStream fsinput = fs.open(p); final DataInputStream input = getWalStream(p, fsinput)) {
```
nit: would prefer to see these split up onto two lines. Not sure what others think :)

adamjshook replied:
The formatter in the build process reformats it so they are on the same line.

ctubbsii replied:
That might be configurable. Would have to investigate the options in newer Eclipse versions. It doesn't matter to me... if it's configurable, I'll defer to others as to what is most readable.

joshelser commented on server/tserver/src/main/java/org/apache/accumulo/tserver/replication/AccumuloReplicaSystem.java:
```diff
@@ -370,11 +368,112 @@ protected Status replicateLogs(ClientContext peerContext, final HostAndPort peer
+        span.stop();
+      }
+
+      log.debug("Sending batches of data to peer tserver");
```
Yuck, this should really be split up into its own method. I know you just copy-pasted, but if you want to come back in here later and clean up some more ;)

adamjshook replied:
Yeah, it's a little nasty.  These are all indentation changes with the addition of the `try-catch`.

keith-turner replied:
Would be nice to open an issue for follow up work.

ctubbsii commented on server/tserver/src/main/java/org/apache/accumulo/tserver/log/DfsLogger.java:
```diff
@@ -414,7 +413,7 @@ public static DFSLoggerInputStreams readHeaderAndReturnStream(VolumeManager fs,
       }
     } catch (EOFException e) {
-      log.warn("Got EOFException trying to read WAL header information, assuming the rest of the file (" + path + ") has no data.");
+      log.warn("Got EOFException trying to read WAL header information, assuming the rest of the file has no data.");
```
I'm not a fan of both logging *and* throwing. I think it's generally better to pick one or the other: handle an exception (by logging in this case) or pass it up. Dropping the path makes it even more obvious that this might be a redundant message... since my first instinct was to suggest adding `, e` to the `log.warn` method call, but that's probably not needed if the `LogHeaderIncompleteException` is keeping the stack trace for logging later.

It's not a big deal... but something to think about if there's an obvious cleanup here.

ctubbsii commented on server/tserver/src/main/java/org/apache/accumulo/tserver/log/LogSorter.java:
```diff
@@ -113,44 +113,46 @@ public void sort(String name, Path srcPath, String destPath) {
+        try (final FSDataInputStream fsinput = fs.open(srcPath)) {
+          DFSLoggerInputStreams inputStreams;
+          try {
+            inputStreams = DfsLogger.readHeaderAndReturnStream(fsinput, conf);
```
I'm not exactly sure what kind of additional resources the crypto stream might use, but it seems to me that since `inputStreams` is not `Closeable`, it may be possible for an the `fsinput` to be closed, but the crypto stream in `inputStreams` to still be holding resources. Would need to dig further to be sure, but it would be nice if `DFSLoggerInputStreams` was `Closeable` and this was created as a second resource, in the `try-with-resources` block on line 116.

Same comment applies to other uses of `DFSLoggerInputStreams`.

adamjshook commented:
Is there a particular process for merging in fixes and applying them to other release lines?

adamjshook commented:
Created [ACCUMULO-4815](https://issues.apache.org/jira/browse/ACCUMULO-4815) to track some cleanup efforts.

joshelser commented:
> Is there a particular process for merging in fixes and applying them to other release lines?

No hard+fast rules. General recommendation is to apply it to the oldest branch and then merge it through newer release lines.

adamjshook commented:
@joshelser Thanks!  Looks like I don't have write access here on Github.  Is that a mistake or should I apply it via `gitbox.apache.org`.

joshelser commented:
Try  following https://gitbox.apache.org/setup/ @adamjshook

ctubbsii commented:
Created follow-on issue for the potentially unclosed crypto stream from `DfsLogger.readHeaderAndReturnStream` https://issues.apache.org/jira/browse/ACCUMULO-4816
